### PR TITLE
ref(avatars): remove Avatar.js

### PR DESCRIPTION
**Pull request by virtuacoplenny**
- Rely on redux getting updated with new participant state
and any calls to getAvatarURL passing in the redux
participant state. This way the state within Avatar.js can
be removed.
- Clean up methods on UI.js. Because all state is in the
store, separate methods for updating the avatar aren't as
necessary. Instead centralize accessing of the avatar for
components outside of redux and centralize the call to
update avatars for non-react components.
- Controversial: cache a participant's avatarURL on the
participant state. Currently the participant's avatarURL
that is generated without jwt (which sets the avatarURL directly)
is not cached. Without cache, there can be many redundant
calls to APP.API.notifyAvatarChanged.

### CODE REVIEW

**Comment by virtuacoplenny** [comment]
@hristoterezov, I think this implementation follows what was discussed in person. I think it raises some concerns, specifically having the two avatarURL values (jwt and locally generated) might not be best for long term clarity. There are some other implementations I'll think about but wanted your updated opinion on this.

**Comment by virtuacoplenny** [comment]
@yanas knows about Avatar flows as well so I'm asking for her review and maybe she has other ideas. The main issue I've been trying to get around with killing Avatar.js is making sure APP.API. notifyAvatarChanged doesn't get called more than it needs to. In current master, Avatar.js has a cache that'll prevent that from happening but killing Avatar.js could mean killing that cache as well. I can't think of a good way to get rid of that cache while minimizing changes to existing code.

**Comment by virtuacoplenny** [comment]
Other ideas I've been tossing around:
1. Keep avatar.js but only as a cache and it gets called through APP.UI whenever a email/name/avatar gets updated, much like what happens now in master.
2. Keep the avatar.js cache somewhere...maybe in UI(?)...but remove the avatar.js file. This isn't really different from keeping avatar.js.
3. Add a redux store subscriber that would listen to whenever the store updates and diff avatarURLs.
4. Add an avatar cache to API.js. Hristo recommended this one; it didn't even come to me as an option until he mentioned it.
5. Keep Avatar.js. Killing Avatar.js was something that seemed right to do because it has cached state outside of redux. If the atlassian avatar initial service is to be used, Avatar.js could be extended to also cache display names so it can pass a display name into getAvatarURL.
6. I could also do something crazier and have middleware that listens for participant updates but calls next(action) AFTER the store has updated. So inside the middleware I could generate and cache the avatar url before the update, then generate the avatar url after the update, and emit to APP.API if there is a difference. I'm not sure how I feel about this use of middleware because I haven't done something like that before in other apps, although that pattern already exists for audio only.

I don't like 6 but I'm leaning towards it now.

**Comment by yanas** [comment]
I'm not sure I understood option 3 btw, can you elaborate on that one? I think it will be a mistake to keep avatar.js and I like the main approach here, so all other options are interesting :)

**Comment by virtuacoplenny** [comment]
Redux has a concept of subscribers, which in short is like a pub-sub to notify subscribers of state updates. It's not unlike how a component can be connected to redux. It's one way to react to state changes and trigger further changes.

File: modules/UI/UI.js [hunk 1]
```diff
@@ -789,26 +790,23 @@ UI.showToolbar = timeout => APP.store.dispatch(showToolbox(timeout));
...
+    const state = APP.store.getState();
+    let participant;
+
+    if (!id || APP.conference.isLocalId(id) || id === 'local') {
+        participant = getLocalParticipant(state);
+    } else {
+        participant = getParticipantById(state, id);
+    }
+
+    return participant ? getAvatarURL(participant) : '';
```
[the remark this answers is not in the document]
Comment (virtuacoplenny) [reply modules/UI/UI.js]: Are you referring to a "selector" where state gets passed in along with a user id, and the selector function returns the avatar url?

I can do that. How about I make a commit to support option 6 and make another commit (in this pr) to create the selector.

[the remark this answers is not in the document]
Comment (yanas) [reply modules/UI/UI.js]: Sounds good.

**Comment by virtuacoplenny** [comment]
Okay, I think I've covered changing the timing in the middleware for diffing avatarURLs, removing as much of refreshAvatarDisplay as I can at this point and moving the avatarURL selector into participant/functions. I had to keep supporting a hack where the local participant id is hardcoded to "local" in redux but is left alone as undefined in conference.js. Please take a look and see if this approach feels better. I'm still not a fan of how the middleware essentially has before and after-update logic but maybe I can get used to it if it doesn't pose long-term problems.

**Pull request merged**